opensquirrel/register_manager.py

```diff
@@ -1,24 +1,142 @@
+        index_to_variable_name: dict[int, str] | None = None,
+    ) -> None:
+        self.register_size: int = register_size
+        self.variable_name_to_range: dict[str, Range] = variable_name_to_range or dict()
+        self.index_to_variable_name: dict[int, str] = index_to_variable_name or dict()
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, Register):
+            return False
+        return (
+            self.register_size == other.register_size
+            and self.variable_name_to_range == other.variable_name_to_range
+            and self.index_to_variable_name == other.index_to_variable_name
+        )
+
+    def get_variable_name(self, index: int) -> str:
+        """Get the variable name at `index`."""
+        return self.index_to_variable_name[index]
+
+    def get_range(self, variable_name: str) -> Range:
+        """Get the Range for a `variable_name`."""
+        return self.variable_name_to_range[variable_name]
+
+    def get_index(self, variable_name: str, sub_index: int) -> int:
+        """Get the Index for a given `subIndex` of a `variable_name`."""
+        return self.variable_name_to_range[variable_name].first + sub_index
+
+    def __repr__(self) -> str:
+        entries: str = ""
+        first: bool = True
+        for variable_name, register_range in self.variable_name_to_range.items():
+            entries += "{}{}: {}".format("" if first else ", ", variable_name, register_range)
+            first = False
+        return f"{{ {entries} }}"
+
+    def size(self) -> int:
+        return self.register_size
+
+    @staticmethod
+    def _parse_ast_string(string: str) -> str:
+        # FIXME: libqasm should return bytes, not the __repr__ of a bytes object ("b'q'")
+        return string[2:-1]
+
+    @classmethod
+    def from_ast(cls, ast: cqasm.semantic.Program) -> Self:
+        register_size: int = 0
+        variable_name_to_range: dict[str, Range] = dict()
+        index_to_variable_name: dict[int, str] = dict()
+
+        variables = [v for v in ast.variables if cls.is_of_type(v)]
+        register_size = sum([v.typ.size for v in variables])
+        current_index: int = 0
+        for v in variables:
+            v_name = Register._parse_ast_string(v.name)
+            v_size = v.typ.size
+            variable_name_to_range[v_name] = Range(current_index, v_size)
+            for _ in range(v_size):
+                index_to_variable_name[current_index] = v_name
+                current_index += 1
+        return cls(register_size, variable_name_to_range, index_to_variable_name)
+
+
+# TODO:
+# In the future, when variables of different types can be defined (e.g. float q)
+# we will have to prevent a variable being called 'q' or 'b'.
+# A possible way to do this would be to introduce variable name mangling in the IR
+# (e.g., store float 'q' as '_float__q')
+# This variable name mangling could be done as an early compiler pass
```

Comment:
What happens now when I define a qubit register 'b' and a bit register 'q'?

Reply:
So this would be the scenario now:
- You create a correct program with a qubit `b` and a bit `q`.
- OpenSquirrel parses the semantic AST coming from libqasm and creates a qubit register `q` and a bit register `b` with the needed amount of qubits and bits respectively.
- The qubit `b` and the bit `q` variables are no longer used. All the code in OpenSquirrel works directly with the qubit register `q` and the bit register `b` through the RegisterManager. For example, a gate such as `X b`, operating on qubit variable `b` will be internally processed as an `X` gate with a `Qubit` operand with a  given index, and that index is a position in the qubit register `q`.
- The RegisterManager still keeps data structures though connecting variable names to ranges in the qubit/bit registers. For example, it could tell you that qubit  variable `b` sits in index 5 of the qubit register `q`.

Reply:
Very good question though!

Reply:
Good to know that it doesn't break. Should we add a unit test with this scenario?

Reply:
Good idea, I'll add one!

Reply:
I have added `test_qubit_variable_b_and_bit_variable_q`.